modules/lily-http/src/main/java/io/github/tomboyo/lily/http/UriTemplate.java

```diff
@@ -1,55 +1,119 @@
 package io.github.tomboyo.lily.http;
 
-import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNullElse;
 
+import io.github.tomboyo.lily.http.encoding.Encoder;
 import java.net.URI;
-import java.net.URLEncoder;
-import java.util.Arrays;
 import java.util.HashMap;
 import java.util.regex.Pattern;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
+/**
+ * A utility that creates URIs from template strings and parameter bindings.
+ *
+ * <p>
+ *
+ * <pre>
+ *   UriTemplate
+ *     .of("https://example.com/", "{myParam}", "{query}{continuation}")
+ *     .bind("myParam", "some;value")
+ *     .bind("query", Map.of("key1", "value1"), Encoding.form(EXPLODE))
+ *     .bind("continuation", List.of("a", "b"), Encoding.formContinuation(EXPLODE))
+ *     .toURI()
+ *     .toString();
+ *     // => https://example.com/some;value?/?key1=value1&continuation=a&continuation=b
+ * </pre>
+ */
 public class UriTemplate {
 
   private final String template;
```

Comment:
Consider making this mutable. The bind API is mutable, whereas the template APIs are not, so the interface isn't consistently mutable or immutable, which might confuse a user.